llama-index-core/llama_index/core/agent/runner/base.py

```diff
@@ -365,7 +365,8 @@ def _run_step(
         **kwargs: Any,
     ) -> TaskStepOutput:
         """Execute step."""
-        dispatcher.event(AgentRunStepStartEvent())
+        span_id = dispatcher.current_span_id
+        dispatcher.event(AgentRunStepStartEvent(span_id=span_id))
```

Comment:
rather than getting the span-id and passing it in, could `event()` attach this automatically?

Reply:
(would make development easier and less error-prone)

Reply:
This is actually why the callback manager used a context manager 
`with callback_manager.event() as event:`

So if the dispatcher needs a similar pattern, we could use that again, or if possible, it just gets handled automatically

Reply:
That's a good point. We could maybe try to adopt a context manager approach:

```
with dispatcher.span_session() as dispatch_event:
     dispatch_event()
     # perform other logic
     dispatch_event()
```

So, here I think we make `dispatcher.event(span_id, event)` and so span_id will added to the event within `event()`. And then the context manager returns a partial of `dispatcher.event(span_id, event)` i.e., `dispatcher_event = partial(dispatcher.event, span_id=dispatcher.current_span_id)`

```
# dispatcher.py
def event(event: BaseEvent, span_id: str = ""):
     # modify span_id field of event
     event.span_id = span_id
     ...
```

Reply:
Since we already decorated the function, I wonder if we can do something like this
https://stackoverflow.com/questions/12435765/can-a-decorated-function-access-variables-of-the-decorator

It would completely hide this from the implementation (and maintains a cleaner look to the code for readability)

Thoughts?

Reply:
This way, when .event() is called, it can automatically pull the span ID from globals? Maybe? (if the injected global is a function/lambda?)

Reply:
If that doesn't work, then context manager it is (even though its slightly less neat)

Reply:
For visibility (lol we're discussing via slack on this too -- my b):

I think this will work! Attach the attr to the wrapper which should still be coroutine safe I think. Will give it a shot.

Reply:
alright, `wrapt` complicated things for me a bit. Resorted to using context manager approach as described previously in this thread.